### PR TITLE
[layers-panel] Changes to layer mark in layer tree view

### LINKED ISSUE #32029: layers panel shows strange artifact
**Describe the bug**

There's a strange grey artifact on the Layers panel.

image

First I thought it might be some new feature but it kinda shows also mixed with other layer symbols.

**How to Reproduce**

Just open some layer on QGIS.

**QGIS and OS versions**

QGIS version | 3.9.0-Master | QGIS code revision | 880de7f364
-- | -- | -- | --
Compiled against Qt | 5.9.5 | Running against Qt | 5.9.5
Compiled against GDAL/OGR | 2.2.3 | Running against GDAL/OGR | 2.2.3
Compiled against GEOS | 3.6.2-CAPI-1.10.2 | Running against GEOS | 3.6.2-CAPI-1.10.2 4d2925d6
Compiled against SQLite | 3.22.0 | Running against SQLite | 3.22.0
PostgreSQL Client Version | 11.5 (Ubuntu 11.5-1.pgdg18.04+1) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.2
PROJ.4 Version | 493
OS Version | Ubuntu 18.04.3 LTS

**Additional context**

### PR DESCRIPTION
**Pull request by gacarrillor**
## Description

PR to suggest 2 changes to layer mark in layers panel:

1. Update the viewport when scrolling horizontally to avoid the flicker.

2. Avoid showing the layer mark always and instead show it only when a layer/group icon is not visible anymore. 

**Before / After** 

before      ..........          after

**Before:** 
image 

**After:**
image

There are other options for suggestion number 2, like showing the layer mark: 
 
2.1 only when the horizontal scroll value is greater than the minimum (0); or
2.2 only when the layer/group text (including feature count) is not visible anymore. This one is a bit trickier!

Fix #32029

### CODE REVIEW

**Comment by nyalldawson** [comment]
@dakcarto?

File: src/gui/layertree/qgslayertreeview.h [hunk 1]
```diff
@@ -234,6 +234,8 @@ class GUI_EXPORT QgsLayerTreeView : public QTreeView
     void onCurrentChanged();
     void onExpandedChanged( QgsLayerTreeNode *node, bool expanded );
     void onModelReset();
+    //! Handles updating the viewport to avoid flicker
+    void onHorizontalScroll( int value );
```
Comment (3nids) [inline src/gui/layertree/qgslayertreeview.h:238]: I think this should be in the private slots

**Review by 3nids — COMMENTED** [review]
1. nice!

2. I am not 100% due to misalignment then. No strong opinion here. Might be worth 2 distinct PRs.

**Comment by gacarrillor** [comment]
@3nids Well, we could also implement it in this way to avoid misalignment: 

no_misalignment

File: src/gui/layertree/qgslayertreeview.h (hunk 1 above)
Comment (gacarrillor) [reply src/gui/layertree/qgslayertreeview.h]: Fixed in https://github.com/qgis/QGIS/pull/35461/commits/6d1b8d6d063084e12168778f4b50e1df4c685c9d

**Comment by gacarrillor** [comment]
Just rebased and applied @3nids comments.

**Review by m-kuhn — COMMENTED** [review]
I'd like to reopen this one.
Resolving the graphical glitches is a good thing, no questions asked. About the layer mark I am not completely sure. To be honest only since I looked at this pull request right now (for about the fifth time) I think I really understood what it means :D -- and with "it" I refer to the layer mark, not the change in this PR.

**Comment by gacarrillor** [comment]
Yes, the layer mark was not the easiest thing to understand for me neither, but I personally think it's a good idea, once all concerns from #32029 are addressed, which this PR does.

Just let me know how should we proceed (I'll apply the suggestion you already left above). 

It would be great if @nirvn, @nyalldawson and @dakcarto can give us an opinion on this PR, since they were involved in #30227, which this PR is attempting to polish.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**